primers/pcr/pcr.go

```diff
@@ -0,0 +1,184 @@
+*/
+package pcr
+
+import (
+	"errors"
+	"github.com/TimothyStiles/poly/primers"
+	"github.com/TimothyStiles/poly/transform"
+	"index/suffixarray"
+	"sort"
+	"strings"
+)
+
+// https://doi.org/10.1089/dna.1994.13.75
+var minimalPrimerLength int = 15
+
+// DesignPrimersWithOverhangs designs two primers to amplify a target sequence,
+// adding on an overhang to the forward and reverse strand. This overhang can
+// contain additional DNA needed for assembly, like Gibson assembly overhangs
+// or GoldenGate restriction enzyme sites.
+func DesignPrimersWithOverhangs(sequence, forwardOverhang, reverseOverhang string, targetTm float64) (string, string) {
+	sequence = strings.ToUpper(sequence)
+	forwardPrimer := sequence[0:minimalPrimerLength]
+	for i := 0; primers.MeltingTemp(forwardPrimer) < targetTm; i++ {
+		forwardPrimer = sequence[0 : minimalPrimerLength+i]
+	}
+	reversePrimer := transform.ReverseComplement(sequence[len(sequence)-minimalPrimerLength:])
+	for i := 0; primers.MeltingTemp(reversePrimer) < targetTm; i++ {
+		reversePrimer = transform.ReverseComplement(sequence[len(sequence)-(minimalPrimerLength+i):])
+	}
+
+	// Add overhangs to primer
+	forwardPrimer = forwardOverhang + forwardPrimer
+	reversePrimer = transform.ReverseComplement(reverseOverhang) + reversePrimer
+
+	return forwardPrimer, reversePrimer
+}
+
+// DesignPrimers designs two primers to amplify a target sequence and only that
+// target sequence (no overhangs).
+func DesignPrimers(sequence string, targetTm float64) (string, string) {
+	return DesignPrimersWithOverhangs(sequence, "", "", targetTm)
+}
+
+// SimulateSimple simulates a PCR reaction. It takes in a list of sequences and
+// a list of primers, with support for complex multiplex reactions, produces
+// a list of all possible PCR fragments from such a reaction. It does not
+// detect concatemerization, which could be useful or very detrimental to
+// your reactions. The variable `circular` is for if the target template is
+// circular, like a plasmid.
+func SimulateSimple(sequences []string, targetTm float64, circular bool, primerList []string) []string {
+	// Set all primers to uppercase.
+	for i := range primerList {
+		primerList[i] = strings.ToUpper(primerList[i])
+	}
+
+	var pcrFragments []string
+	for _, sequence := range sequences {
+		sequence = strings.ToUpper(sequence)
+		sequenceIndex := suffixarray.New([]byte(sequence))
+
+		primerLength := len(primerList)
+
+		forwardLocations := make(map[int][]int)
+		reverseLocations := make(map[int][]int)
+		minimalPrimers := make([]string, primerLength)
+		for primerIdx, primer := range primerList {
+			var minimalLength int
+			for index := minimalPrimerLength; primers.MeltingTemp(primer[len(primer)-index:]) < targetTm; index++ {
+				minimalLength = index
+				if primer[len(primer)-index:] == primer {
+					break
+				}
+			}
+			// Use the minimal binding sites of the primer to find positions in the template
+			if primer[len(primer)-minimalLength:] != primer {
```

Comment:
redundancy of using `len(primer)-minimalLength`. create minimalPrimer before if already resolve most of the repetititons:
```
minimalPrimer := primer[len(primer)-minimalLength:]
if minimalPrimer  != primer {
```

Reply:
You are 100% right, thank you for pointing this out.